marlowe-runtime/tx/Language/Marlowe/Runtime/Transaction/Safety.hs

```diff
@@ -0,0 +1,389 @@
+  , ProtocolParameters(ProtocolParameters, protocolParamUTxOCostPerByte)
+  , SystemStart(..)
+  )
+import qualified Data.Map.Strict as M (Map, elems, empty, fromList, keys, map, mapKeys, singleton, size)
+import qualified Data.Set as S (singleton)
+import qualified Language.Marlowe.Core.V1.Merkle as V1 (MerkleizedContract(..))
+import qualified Language.Marlowe.Core.V1.Plate as V1 (extractAllWithContinuations)
+import qualified Language.Marlowe.Core.V1.Semantics as V1
+  (MarloweData(..), MarloweParams(..), TransactionInput(..), TransactionOutput(..))
+import qualified Language.Marlowe.Core.V1.Semantics.Types as V1 (Party(Address), State(..), Token(..))
+import qualified Language.Marlowe.Core.V1.Semantics.Types.Address as V1 (deserialiseAddress)
+import qualified Language.Marlowe.Runtime.ChainSync.Api as Chain
+  ( Address(..)
+  , AssetId(..)
+  , Assets(..)
+  , DatumHash(..)
+  , PolicyId(..)
+  , SlotNo(..)
+  , TokenName(..)
+  , Tokens(..)
+  , TransactionMetadata(..)
+  , TransactionOutput(..)
+  , TxOutRef
+  , UTxOs(..)
+  )
+import qualified Ouroboros.Consensus.BlockchainTime as Ouroboros (RelativeTime(..), mkSlotLength, toRelativeTime)
+import qualified Ouroboros.Consensus.HardFork.History as Ouroboros
+  ( Bound(..)
+  , EraEnd(..)
+  , EraParams(..)
+  , EraSummary(..)
+  , SafeZone(..)
+  , interpretQuery
+  , mkInterpreter
+  , summaryWithExactly
+  , wallclockToSlot
+  )
+import qualified Ouroboros.Consensus.Util.Counting as Ouroboros (Exactly(..))
+import qualified Ouroboros.Network.Block as Ouroboros (SlotNo(..))
+import qualified Plutus.V2.Ledger.Api as Plutus
+  (CurrencySymbol(..), DatumHash(..), POSIXTime(..), TokenName(..), fromBuiltin, toBuiltin)
+import qualified PlutusTx.AssocMap as AM (toList)
+
+
+-- FIXME: Relocate this definition when full support for Merkleization is added to Runtime.
+type Continuations v = M.Map Chain.DatumHash (Contract v)
+
+
+-- | No continuations for a contract.
+noContinuations
+  :: MarloweVersion v
+  -> Continuations v
+noContinuations MarloweV1 = M.empty
+
+
+-- | Map Plutus continuations into chain-sync continuations.
+remapContinuations
+  :: M.Map Chain.DatumHash contract
+  -> M.Map Plutus.DatumHash contract
+remapContinuations = M.mapKeys $ Plutus.DatumHash . Plutus.toBuiltin . Chain.unDatumHash
+
+
+-- | Compute a worst-case bound on the minimum UTxO value for a contract.
+minAdaUpperBound
+ :: Shelley.ProtocolParameters
+ -> MarloweVersion v
+ -> Contract v
+ -> Continuations v
+ -> Maybe Cardano.Lovelace
+minAdaUpperBound Shelley.ProtocolParameters{protocolParamUTxOCostPerByte} MarloweV1 contract continuations =
+  fromInteger
+    . (\utxoCostPerByte -> worstMinimumUtxo (toInteger utxoCostPerByte) contract $ remapContinuations continuations)
+    <$> protocolParamUTxOCostPerByte
+
+
+-- | Check a contract for design errors and ledger violations.
+checkContract
+  :: RoleTokensConfig
+  -> MarloweVersion v
+  -> Contract v
+  -> Continuations v
+  -> [SafetyError]
```

Comment:
Could this return the full `SafetyReport`? Then we can make the user-provided min ADA deposit optional and use that directly.

Reply:
That's certainly possible--it sort of depends upon the use case for these function. We may want to consider designing the REST API and then altering this library to conveniently enable that use case.

It is possible that `checkContract` may not be needed because `checkTransactions` will catch everything. However, `checkContract` is far faster than `checkTransactions`, so it might be worth calling the former first, just to catch low-hanging fruit before spending resources on a pointless `checkTransactions`.

Reply:
I think we should wait until we're further along on the REST sides of this because this relates to where we want to perform the protocol-dependent checks, where the min-UTxO is chosen (and how the user might override that), etc.

There are really three phases to the safety checking:
1. Select the protocol-dependent min-UTxO for the contract and the account into which that is placed.
2. Check the protocol-independent aspects of the contract.
3. Run the transactions for all possible execution paths, which is expensive.